meson.build

```diff
@@ -0,0 +1,77 @@
+# SPDX-License-Identifier: LGPL-2.1+
+project('libbpf', 'c',
+        version : '0.0.2',
+        license : 'LGPLv2.1')
+
+libbpf_src_dir = './'
+
+libbpf_sources = files(
+        join_paths(libbpf_src_dir, 'src/bpf.c'),
+        join_paths(libbpf_src_dir, 'src/bpf.h'),
+        join_paths(libbpf_src_dir, 'src/bpf_prog_linfo.c'),
+        join_paths(libbpf_src_dir, 'src/btf.c'),
+        join_paths(libbpf_src_dir, 'src/btf.h'),
+        join_paths(libbpf_src_dir, 'src/libbpf.c'),
+        join_paths(libbpf_src_dir, 'src/libbpf.h'),
+        join_paths(libbpf_src_dir, 'src/libbpf_errno.c'),
+        join_paths(libbpf_src_dir, 'src/libbpf_probes.c'),
+        join_paths(libbpf_src_dir, 'src/libbpf_util.h'),
+        join_paths(libbpf_src_dir, 'src/netlink.c'),
+        join_paths(libbpf_src_dir, 'src/nlattr.c'),
+        join_paths(libbpf_src_dir, 'src/nlattr.h'),
+        join_paths(libbpf_src_dir, 'src/str_error.c'),
+        join_paths(libbpf_src_dir, 'src/str_error.h'),
+        join_paths(libbpf_src_dir, 'src/xsk.c'),
+        join_paths(libbpf_src_dir, 'src/xsk.h'))
```

Comment:
@rdna would it be possible to set up Travis CI or Semaphore CI that would run `meson ./build && ninja -C ./build` to make sure that `libbpf` could be built with `meson`? If there were any tests it could also run `ninja -C ./build test` to be 100% sure that nothing was forgotten.

Reply:
I am not familiar with Travis CI or Semaphore CI. But I think it is a good idea to integrate them if possible to add test coverage. But that is a different subject than this pull request.

Reply:
This PR is supposed to make it possible to build `libbpf` with `meson` and it seems reasonable to me to add something that would help to make sure that it works and it won't be broken in a week because it's easy to forget to handle this file. It just seems to me that the systemd CI isn't the right place to make sure that `libbpf` can be built with `clang` and `-werror` for example.

Reply:
I just realized that it's possible to point `meson` to @wat-ze-hex's repository in https://github.com/systemd/systemd/pull/12151 to get around the issue with submodules on Ubuntu CI (where I just turned on ARM just in case) (that is, there is no need to wait for this PR to be merged). It seems to be easier and faster at this point.

Reply:
@evverx yeah, that's actually very good idea. We can do it either separately, as @yonghong-song suggested, or here, no strong preference.

I'm not familiar with these CI either, but trying to play with Semaphore now to set it up for my fork of libbpf. I see it has many different settings in `semaphore.yml` so if you have any recommendations on how to set it up to keep testing procedure as close as possible to systemd process that'd be really helpful. E.g. I see that by default it uses `ubuntu1804` image -- not sure if it's good enough for our first step.  We can start with something simple and extend it along the way.

Reply:
> Is there a way to use some masks here (`*.h` / `*.c` ?) like it's done in `src/Makefile`? So that it doesn't go out of sync when new files are added to the library since it's easy to forget to handle this file.

Can't see that Meson has regex builtin. Well, as a last resort we can shell out:
```
libbpf_sources = files(run_command('find',
        ['@0@/src'.format(libbpf_src_dir), '-type',  'f', '-name', '*.h',  '-o',  '-name', '*.c']).stdout().split())
```
Looks ugly but it works.

@evverx
>I just realized that it's possible to point meson to @wat-ze-hex's repository in systemd/systemd#12151 to get around the issue with submodules on Ubuntu CI (where I just turned on ARM just in case) (that is, there is no need to wait for this PR to be merged). It seems to be easier and faster at this point.
I had this idea but then there was another one: that pointing to my fork tastes like anarchy :) I can do that.
Is there way to set up Ubuntu CI just like Travis CI or Semaphore CI ?

Reply:
@rdna judging by `ubuntu1804`, it seems to be Semaphore 2.0, which I've never used unfortunately. We use "classic" Semaphore, where we run debian-testing containers on top of Ubuntu 16.04, which is probably overkill given that you already have access to a more or less fresh build environment. I'll try to take a look at the documentation to figure out how to install `gcc`, `clang`, `libelf`, `meson` and `ninja` and run `meson && ninja` there.

@wat-ze-hex I agree pointing `meson` to your repository would have been a bit chaotic (in a good way) but it would help to keep experimenting with `libbpf` in that PR without going through the libbpf repository where you need to think about the licence headers and other boring and time-consuming stuff. That I think could be done later (including the CI part on the libbpf side).

Regarding Ubuntu CI, it was set up by the debian and ubuntu maintainers a long time ago. I'm not saying it isn' possible but I'm afraid I can't be of any help here.

Reply:
@evverx thank you for all the details! Just to double-check .. do you recommend to go ahead with Semaphore 2.0 and this `ubuntu1804` image? It seems this process may take some time due to learning curve so any help on this is appreciated :)

Reply:
@rdna I'm looking at the "pricing" page there and it seems to me that Semaphore 2.0 isn't exactly free for open-source project (though, they accept all the credit cards, which is reassuring :-)). I'll try to dig deeper later but right now I can't say I'd choose it.

In the meantime, I took a couple of scripts from `systemd` and changed them so that they could build `libbpf` only (with `gcc`, `clang` and also with ASAN/UBsan but I guess that stages can be safely removed because they don't make much sense without tests). Judging by https://github.com/evverx/libbpf/pull/1, it's more or less working and in theory if Travis CI was turned on, it would be possible to add [the second commit](https://github.com/evverx/libbpf/pull/1/commits/fff6f1df402f23e23c3c5541fa648abd020dad85) (or its variation) here to get the ball rolling.

Reply:
> I just realized that it's possible to point `meson` to @wat-ze-hex's repository in [systemd/systemd#12151](https://github.com/systemd/systemd/pull/12151) to get around the issue with submodules on Ubuntu CI (where I just turned on ARM just in case) (that is, there is no need to wait for this PR to be merged). It seems to be easier and faster at this point.

Done, still getting failures. The problem is for sure that it's explicitly forbidden to download wraps in Ubuntu CI, they have `--wrap-mode=nodownload`
https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-bionic-upstream-systemd-ci-systemd-ci/bionic/arm64/s/systemd-upstream/20190411_035033_c0833@/log.gz

So, from this point it's only fixing the tests.

Also LGTM failed. Think due to `--recurse-submodules=no` is on, but I'm not sure.

Reply:
> The problem is for sure that it's explicitly forbidden to download wraps in Ubuntu CI, they have --wrap-mode=nodownload

@wat-ze-hex I didn't know that. That means that it doesn't seem to be possible to do anything without the maintainers of Ubuntu CI (and I already CCed them in that PR). So it appears the switch wasn't necessary. I'm sorry about that.

Regrading LGTM, to judge from https://discuss.lgtm.com/t/lgtm-seems-to-be-failing-to-extract-the-systemd-project-due-to-a-submodule/1994/4?u=evverx they use some kind of cache that should be updated manually but I guess it would be better to wait until everything else is fixed before asking them to update the cache. Plus I turned LGTM on in https://github.com/evverx/libbpf/pull/1 and it didn't find anything so the alerts are the same: https://lgtm.com/projects/g/evverx/libbpf/alerts/?mode=tree.

I'm not sure why the tests are still failing under UBSan though. In https://github.com/libbpf/libbpf/issues/28 I applied the patch and the issue wasn't there anymore. Is that patch included in the branch that was tested under UBsan?

Reply:
>So it appears the switch wasn't necessary. I'm sorry about that.

No problem, that's the right approach so libbpf maintainers can keep `meson.build` in sync with Makefile. Also if meson provides capabilities to manage subprojects, we should utilize them.

> I'm not sure why the tests are still failing under UBSan though. In #28 I applied the patch and the issue wasn't there anymore. Is that patch included in the branch that was tested under UBsan?

My `master` doesn't have the fix. Waiting for https://patchwork.ozlabs.org/patch/1083690/ in libbpf upstream.